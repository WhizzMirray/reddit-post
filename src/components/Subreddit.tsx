import * as React from "react";

import Post from "./Post";

import { IPost, IPostBySubreddit, ISettings } from "../types/interfaces";

export interface IProps {
  reddit: IPostBySubreddit;
  title: string;
  settings: ISettings;
  onReload: any;
  onRemove: any;
  callNotification?: any;
  playAudio?: any;
}
interface IState {
  collapsed: boolean;
  openedNewPost: boolean;
}


class Subreddit extends React.Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);   
    this.state = {
      collapsed: true,
      openedNewPost: false,
    };
    this.handleClick = this.handleClick.bind(this);
  }

  public componentWillReceiveProps(nextProps: IProps) {
    const { playAudio, callNotification, settings} = nextProps;
    if (nextProps.reddit.hasNewPost) {
      this.setState({
        openedNewPost: false
      });

      if (settings.notification && callNotification) {
        callNotification();
      }
      
      if (settings.sound && playAudio) {
        playAudio();
      }
    }
  }

  public render() {
    const posts = this.props.reddit.items.map((item: IPost) => (
      <Post key={item.id} {...item} />
    ));
    const { isFetching, error, hasNewPost } = this.props.reddit;
    const { onReload, onRemove } = this.props;
    const { openedNewPost } = this.state;
    return (
      <div className="panel panel-default">
        <div className="panel-heading">
          <h4 className="panel-title">
            <a onClick={this.handleClick} style={{ cursor: "pointer" }}>
              {this.props.title}
            </a>
          </h4>

          <span className="pull-right clickable">
            <button
              className="btn btn-default"
              type="button"
              onClick={onReload}
            >
              <span className="glyphicon glyphicon-refresh" />
            </button>

            <button
              className="btn btn-default"
              type="button"
              onClick={onRemove}
            >
              <span className="glyphicon glyphicon-remove" />
            </button>
          </span>

          <span className="time-span pull-right" style={{ marginRight: "5px" }}>
            {isFetching || error
              ? ""
              : "Last fetched : " +
                new Date(this.props.reddit.lastUpdated).toLocaleTimeString()}
          </span>

          <span className="info-span pull-right">
            {isFetching || error || !hasNewPost || openedNewPost
              ? ""
              : "New Post !"}
          </span>
        </div>
        <div
          id="collapse1"
          className={
            "panel-collapse collapse " + (this.state.collapsed ? "" : "in")
          }
        >
          <div className="panel-body">
            {isFetching
              ? "Fetching posts..."
              : error
                ? "Can't fetch posts"
                : posts}
          </div>
        </div>
      </div>
    );
  }

  public handleClick() {
    this.setState(prevState => ({
      collapsed: !prevState.collapsed,
      openedNewPost: true
    }));
  }
}

export default Subreddit;
