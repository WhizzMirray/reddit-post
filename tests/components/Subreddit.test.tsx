import { shallow } from "enzyme";
import * as React from "react";

import Subreddit, { IProps } from "../../src/components/Subreddit";

const makeProps = (): IProps => ({
  callNotification: jest.fn(),
  onReload: jest.fn(),
  onRemove: jest.fn(),
  playAudio: jest.fn(),
  reddit: {
    didInvalidate: false,
    error: false,
    hasNewPost: false,
    isFetching: false,
    items: [],
    lastUpdated: 0
  },
  settings: {
    delay: 5,
    notification: false,
    sound: false
  },
  title: ""
});

describe("<Subreddit/>", () => {
  it("should render correctly", () => {
    const props = makeProps();
    const wrapper = shallow(<Subreddit {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should change the state when the <a> containing the title is clicked", () => {
    const props = makeProps();
    const wrapper = shallow(<Subreddit {...props} />);
    const titleA = wrapper.find("#title-a");

    titleA.simulate("click");
    expect(wrapper.state()).toEqual({
      collapsed: false,
      openedNewPost: true
    });

    titleA.simulate("click");
    expect(wrapper.state()).toEqual({
      collapsed: true,
      openedNewPost: true
    });
  });
  
  it("should change the state, callNotification and playAudio when new props arrive with new posts", () => {
    const props = makeProps();
    Object.assign(props.settings, { notification: true, sound: true });
    const wrapper = shallow(<Subreddit {...props} />);
    wrapper.setState({
      openedNewPost: true
    });

    wrapper.setProps({
      reddit: {
        didInvalidate: false,
        error: false,
        hasNewPost: true, // we check if this changed
        isFetching: false,
        items: [],
        lastUpdated: 0
      }
    });

    expect(wrapper.state()).toEqual({
      collapsed: true,
      openedNewPost: false
    });
    expect(props.callNotification).toHaveBeenCalledTimes(1);
    expect(props.playAudio).toHaveBeenCalledTimes(1);
  });

  it("should call onReload when button is clicked", () => {
    const props = makeProps();
    const wrapper = shallow(<Subreddit {...props} />);
    wrapper.find("#reload-button").simulate("click");

    expect(props.onReload).toHaveBeenCalledTimes(1);
  });

  it("should call onRemove when button is clicked", () => {
    const props = makeProps();
    const wrapper = shallow(<Subreddit {...props} />);
    wrapper.find("#remove-button").simulate("click");

    expect(props.onRemove).toHaveBeenCalledTimes(1);
  });
});
