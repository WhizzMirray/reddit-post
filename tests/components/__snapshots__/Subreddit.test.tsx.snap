// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Subreddit/> should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Subreddit
    callNotification={[MockFunction]}
    onReload={[MockFunction]}
    onRemove={[MockFunction]}
    playAudio={[MockFunction]}
    reddit={
      Object {
        "didInvalidate": false,
        "error": false,
        "hasNewPost": false,
        "isFetching": false,
        "items": Array [],
        "lastUpdated": 0,
      }
    }
    settings={
      Object {
        "delay": 5,
        "notification": false,
        "sound": false,
      }
    }
    title=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="panel-heading"
        >
          <h4
            className="panel-title"
          >
            <a
              id="title-a"
              onClick={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
            >
              
            </a>
          </h4>
          <span
            className="pull-right clickable"
          >
            <button
              className="btn btn-default"
              id="reload-button"
              onClick={[MockFunction]}
              type="button"
            >
              <span
                className="glyphicon glyphicon-refresh"
              />
            </button>
            <button
              className="btn btn-default"
              id="remove-button"
              onClick={[MockFunction]}
              type="button"
            >
              <span
                className="glyphicon glyphicon-remove"
              />
            </button>
          </span>
          <span
            className="time-span pull-right"
            id="last-fetched-span"
            style={
              Object {
                "marginRight": "5px",
              }
            }
          >
            Last fetched : 00:00:00
          </span>
          <span
            className="info-span pull-right"
            id="new-post-span"
          >
            
          </span>
        </div>,
        <div
          className="panel-collapse collapse "
          id="collapse-div"
        >
          <div
            className="panel-body"
            id="posts-div"
          />
        </div>,
      ],
      "className": "panel panel-default",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h4
              className="panel-title"
            >
              <a
                id="title-a"
                onClick={[Function]}
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
              >
                
              </a>
            </h4>,
            <span
              className="pull-right clickable"
            >
              <button
                className="btn btn-default"
                id="reload-button"
                onClick={[MockFunction]}
                type="button"
              >
                <span
                  className="glyphicon glyphicon-refresh"
                />
              </button>
              <button
                className="btn btn-default"
                id="remove-button"
                onClick={[MockFunction]}
                type="button"
              >
                <span
                  className="glyphicon glyphicon-remove"
                />
              </button>
            </span>,
            <span
              className="time-span pull-right"
              id="last-fetched-span"
              style={
                Object {
                  "marginRight": "5px",
                }
              }
            >
              Last fetched : 00:00:00
            </span>,
            <span
              className="info-span pull-right"
              id="new-post-span"
            >
              
            </span>,
          ],
          "className": "panel-heading",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <a
                id="title-a"
                onClick={[Function]}
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
              >
                
              </a>,
              "className": "panel-title",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "",
                "id": "title-a",
                "onClick": [Function],
                "style": Object {
                  "cursor": "pointer",
                },
              },
              "ref": null,
              "rendered": "",
              "type": "a",
            },
            "type": "h4",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <button
                  className="btn btn-default"
                  id="reload-button"
                  onClick={[MockFunction]}
                  type="button"
                >
                  <span
                    className="glyphicon glyphicon-refresh"
                  />
                </button>,
                <button
                  className="btn btn-default"
                  id="remove-button"
                  onClick={[MockFunction]}
                  type="button"
                >
                  <span
                    className="glyphicon glyphicon-remove"
                  />
                </button>,
              ],
              "className": "pull-right clickable",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <span
                    className="glyphicon glyphicon-refresh"
                  />,
                  "className": "btn btn-default",
                  "id": "reload-button",
                  "onClick": [MockFunction],
                  "type": "button",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "glyphicon glyphicon-refresh",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "span",
                },
                "type": "button",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <span
                    className="glyphicon glyphicon-remove"
                  />,
                  "className": "btn btn-default",
                  "id": "remove-button",
                  "onClick": [MockFunction],
                  "type": "button",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "glyphicon glyphicon-remove",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "span",
                },
                "type": "button",
              },
            ],
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Last fetched : 00:00:00",
              "className": "time-span pull-right",
              "id": "last-fetched-span",
              "style": Object {
                "marginRight": "5px",
              },
            },
            "ref": null,
            "rendered": "Last fetched : 00:00:00",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "",
              "className": "info-span pull-right",
              "id": "new-post-span",
            },
            "ref": null,
            "rendered": "",
            "type": "span",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="panel-body"
            id="posts-div"
          />,
          "className": "panel-collapse collapse ",
          "id": "collapse-div",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "panel-body",
            "id": "posts-div",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="panel-heading"
          >
            <h4
              className="panel-title"
            >
              <a
                id="title-a"
                onClick={[Function]}
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
              >
                
              </a>
            </h4>
            <span
              className="pull-right clickable"
            >
              <button
                className="btn btn-default"
                id="reload-button"
                onClick={[MockFunction]}
                type="button"
              >
                <span
                  className="glyphicon glyphicon-refresh"
                />
              </button>
              <button
                className="btn btn-default"
                id="remove-button"
                onClick={[MockFunction]}
                type="button"
              >
                <span
                  className="glyphicon glyphicon-remove"
                />
              </button>
            </span>
            <span
              className="time-span pull-right"
              id="last-fetched-span"
              style={
                Object {
                  "marginRight": "5px",
                }
              }
            >
              Last fetched : 00:00:00
            </span>
            <span
              className="info-span pull-right"
              id="new-post-span"
            >
              
            </span>
          </div>,
          <div
            className="panel-collapse collapse "
            id="collapse-div"
          >
            <div
              className="panel-body"
              id="posts-div"
            />
          </div>,
        ],
        "className": "panel panel-default",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h4
                className="panel-title"
              >
                <a
                  id="title-a"
                  onClick={[Function]}
                  style={
                    Object {
                      "cursor": "pointer",
                    }
                  }
                >
                  
                </a>
              </h4>,
              <span
                className="pull-right clickable"
              >
                <button
                  className="btn btn-default"
                  id="reload-button"
                  onClick={[MockFunction]}
                  type="button"
                >
                  <span
                    className="glyphicon glyphicon-refresh"
                  />
                </button>
                <button
                  className="btn btn-default"
                  id="remove-button"
                  onClick={[MockFunction]}
                  type="button"
                >
                  <span
                    className="glyphicon glyphicon-remove"
                  />
                </button>
              </span>,
              <span
                className="time-span pull-right"
                id="last-fetched-span"
                style={
                  Object {
                    "marginRight": "5px",
                  }
                }
              >
                Last fetched : 00:00:00
              </span>,
              <span
                className="info-span pull-right"
                id="new-post-span"
              >
                
              </span>,
            ],
            "className": "panel-heading",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <a
                  id="title-a"
                  onClick={[Function]}
                  style={
                    Object {
                      "cursor": "pointer",
                    }
                  }
                >
                  
                </a>,
                "className": "panel-title",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "",
                  "id": "title-a",
                  "onClick": [Function],
                  "style": Object {
                    "cursor": "pointer",
                  },
                },
                "ref": null,
                "rendered": "",
                "type": "a",
              },
              "type": "h4",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <button
                    className="btn btn-default"
                    id="reload-button"
                    onClick={[MockFunction]}
                    type="button"
                  >
                    <span
                      className="glyphicon glyphicon-refresh"
                    />
                  </button>,
                  <button
                    className="btn btn-default"
                    id="remove-button"
                    onClick={[MockFunction]}
                    type="button"
                  >
                    <span
                      className="glyphicon glyphicon-remove"
                    />
                  </button>,
                ],
                "className": "pull-right clickable",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <span
                      className="glyphicon glyphicon-refresh"
                    />,
                    "className": "btn btn-default",
                    "id": "reload-button",
                    "onClick": [MockFunction],
                    "type": "button",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "className": "glyphicon glyphicon-refresh",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "span",
                  },
                  "type": "button",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <span
                      className="glyphicon glyphicon-remove"
                    />,
                    "className": "btn btn-default",
                    "id": "remove-button",
                    "onClick": [MockFunction],
                    "type": "button",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "className": "glyphicon glyphicon-remove",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "span",
                  },
                  "type": "button",
                },
              ],
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Last fetched : 00:00:00",
                "className": "time-span pull-right",
                "id": "last-fetched-span",
                "style": Object {
                  "marginRight": "5px",
                },
              },
              "ref": null,
              "rendered": "Last fetched : 00:00:00",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "",
                "className": "info-span pull-right",
                "id": "new-post-span",
              },
              "ref": null,
              "rendered": "",
              "type": "span",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="panel-body"
              id="posts-div"
            />,
            "className": "panel-collapse collapse ",
            "id": "collapse-div",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [],
              "className": "panel-body",
              "id": "posts-div",
            },
            "ref": null,
            "rendered": Array [],
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "prevContext": true,
          },
        },
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
